dev/tests/integration/testsuite/Magento/Catalog/Block/Product/ListTest.php

```diff
@@ -39,7 +43,7 @@ public function testGetLoadedProductCollection()
     {
         $this->_block->setShowRootCategory(true);
         $collection = $this->_block->getLoadedProductCollection();
-        $this->assertInstanceOf(\Magento\Catalog\Model\ResourceModel\Product\Collection::class, $collection);
+        $this->assertInstanceOf(\Magento\Eav\Model\Entity\Collection\AbstractCollection::class, $collection);
         /* Check that root category was defined for Layer as current */
         $this->assertEquals(2, $this->_block->getLayer()->getCurrentCategory()->getId());
     }
```

Comment:
Could you add one more test case that will make sure that category_id was set before to HTML, maybe using reflection to call before to HTML method?

Reply:
I'm not sure how to coverage that when this method is protected and only can callable in origin class

Reply:
Also i not sure the part previously someone try to set `category_id` to product before render to html. Seem not well-write implements

Reply:
I'm not sure the way to check category_id after `_beforeToHtml` called. This out of scope of my knowledge on this part